Add noxfile

Spectre5 opened the pull request:
This PR comes without any previous discussion, so please feel free to just close this if it is undesired.

This adds a noxfile to automatically run the tests, coverage, and docs.  I've setup the Nox sessions to run essentially the same thing that is done in GitHub actions and the same way that read the docs should be building the docs.  Using Nox was helpful in my last PR.

If you're unfamiliar with Nox, it is similar to tox but uses a Python file to run the various sessions.  So with this, you can just run `nox` to run all tests on each Python version (use pyenv to manage them) and build the documentation locally.

https://nox.thea.codes/en/stable/index.html

I personally use pipx to install and manage nox.

anntzer commented:
Can you rebase the PR to exclude the commits related to you previous PR?  Unfortunately it looks like the manual merge up is confusing the github interface.

Spectre5 commented:
Ok, think I got it fixed

Spectre5 commented:
Pushed two more changes:

- Removed `sphinx-rtd-theme`, which isn't used by the docs
- Added a style session which uses flake8 to check *only* the max line length, which is set to 80 (same as the current github action `style` uses).

Spectre5 commented:
I'm not sure what happened here.  I didn't close this, it isn't merged, and I can't reopen it.

anntzer commented on noxfile.py:
```diff
@@ -0,0 +1,64 @@
+    session.install(
+        'sphinx',
+        'sphinx-autodoc-typehints',
+    )
```
please use doc/requirements.txt (adding sphinx in there) to avoid things going out of sync.

Spectre5 replied:
Done.

anntzer commented on noxfile.py:
```diff
@@ -0,0 +1,61 @@
+
+    try:
+        session.run(
+            'coverage', 'run', '--parallel', '--module', 'unittest', *args
```
On the one hand I see that --parallel is handy, on the other does that mean that you could run into outdated coverage files accidentally getting combined by the combine step?  I don't use nox myself so I have no idea how best to handle this; perhaps something like (inventing API here, but you get the idea)
```
def coverage(...):
    python_versions = ...  # something from the command line?
    for version in python_versions: tests(version, coverage_file=<known_path>)
    session.run('coverage', 'combine', <known_paths>)
```
?

Spectre5 replied:
I've never really had an issue with it, and you could always remove all .coverage files first.  Coverage combine removes all of the files that it combines together, so assume it runs then there would only be a single .coverage file leftover.  And coverage combine always ignores that file and overwrites it the next time.  You could potentially end up with a extra .coverage.* file if the nox session was stopped early or something and then it would be added in the coverage combine the next time it was run.

Also, Nox doesn't actually run in parallel, it runs them all sequentially.  I'm just used to using parallel when I setup stuff in CI.  So we could change this to just use --append and let each subsequent python version just append to the single .coverage file.  Perhaps we would need to first delete the coverage file though?

One option would be to make a separate session that simply removes all coverage files first, then triggers the test session (similar to how the test session triggers the coverage one in the current implementation).

Another option would be to just make it all explicit: remove the parallel flag and define the coverage file for each run using the COVERAGE_FILE environment variable.  This is similar to what you do in GitHub actions currently after re-reviewing it.

anntzer replied:
Let's go for COVERAGE_FILE then.

Spectre5 replied:
Added

Spectre5 commented:
Ok, I've used COVERAGE_FILE in the latest build.  I've also added running with the oldest supported versions (same as those in the `build.yml` file).  This means the tests takes twice as long, but they are fairly quick.  Another option would be to not run the old version tests by default, only if supplied another argument, e.g. `nox --old` or something.  Could use the posargs to check for --old or else skip those tests, or remove the parameterization and have a separate settion "tests_old" that is not run by default.  Then it could be run by executing `nox -s tests_old`.

anntzer reviewed (approved):
LGTM, let's not bother with adding more options for now.
